src/esp/gfx/PbrDrawable.cpp

```diff
@@ -0,0 +1,211 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include "PbrDrawable.h"
+
+#include <Corrade/Containers/ArrayViewStl.h>
+#include <Corrade/Utility/FormatStl.h>
+
+namespace Mn = Magnum;
+
+namespace esp {
+namespace gfx {
+
+PbrDrawable::PbrDrawable(scene::SceneNode& node,
+                         Mn::GL::Mesh& mesh,
+                         gfx::Drawable::Flags& meshAttributeFlags,
+                         ShaderManager& shaderManager,
+                         const Mn::ResourceKey& lightSetupKey,
+                         const Mn::ResourceKey& materialDataKey,
+                         DrawableGroup* group)
+    : Drawable{node, mesh, group},
+      shaderManager_{shaderManager},
+      lightSetup_{shaderManager.get<LightSetup>(lightSetupKey)},
+      materialData_{
+          shaderManager.get<MaterialData, PbrMaterialData>(materialDataKey)} {
+  flags_ = PbrShader::Flag::ObjectId;
+
+  if (materialData_->textureMatrix != Mn::Matrix3{}) {
+    flags_ |= PbrShader::Flag::TextureTransformation;
+  }
+  if (materialData_->baseColorTexture) {
+    flags_ |= PbrShader::Flag::BaseColorTexture;
+  }
+  if (materialData_->occlusionRoughnessMetallicTexture) {
+    flags_ |= PbrShader::Flag::OcclusionRoughnessMetallicTexture;
+  }
+  if (materialData_->noneRoughnessMetallicTexture) {
+    flags_ |= PbrShader::Flag::NoneRoughnessMetallicTexture;
+  }
+  if (materialData_->roughnessTexture) {
+    flags_ |= PbrShader::Flag::RoughnessTexture;
+  }
+  if (materialData_->metallicTexture) {
+    flags_ |= PbrShader::Flag::MetallicTexture;
+  }
+  if (materialData_->normalTexture) {
+    flags_ |= PbrShader::Flag::NormalTexture;
+    if (meshAttributeFlags & gfx::Drawable::Flag::HasTangent) {
+      flags_ |= PbrShader::Flag::PrecomputedTangent;
+    }
+    if (materialData_->normalTextureScale != 1.0f) {
+      flags_ |= PbrShader::Flag::NormalTextureScale;
+      CORRADE_ASSERT(materialData_->normalTextureScale > 0.0f,
+                     "PbrDrawable::PbrDrawable(): the normal texture scale "
+                     "cannot be negative.", );
```

Comment:
I know I'm annoyingly picky here, but

```suggestion
                     "has to be positive.", );
```

:trollface:

Reply:
Not a problem at all. Happy to change it.